In [1]:
!pip install xlrd

## Import dataset

In [2]:
import pandas as pd
df=pd.read_excel('heart_failure_prediction.xlsx')
df.head()

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

## Missing Values

In [4]:
df['age'].isnull().sum() # First column

0

In [5]:
df['anaemia'].isnull().sum() # Second column

0

In [6]:
df['creatinine_phosphokinase'].isnull().sum() # Third column

0

In [7]:
df['diabetes'].isnull().sum() # Fourth column

0

In [8]:
df['ejection_fraction'].isnull().sum() # Fifth column

0

In [9]:
df['high_blood_pressure'].isnull().sum() # Sixth column

0

In [10]:
df['platelets'].isnull().sum() # Seventh column

0

In [11]:
df['serum_creatinine'].isnull().sum() # Eighth column

0

In [12]:
df['serum_sodium'].isnull().sum() # Ninth column

0

In [13]:
df['sex'].isnull().sum() # Tenth column

0

In [14]:
df['smoking'].isnull().sum() # Eleventh column

0

In [15]:
df['time'].isnull().sum() # Twelfth column

0

In [16]:
df['DEATH_EVENT'].isnull().sum() # Output column

0

## Feature Selection

In [3]:
import numpy as np

In [4]:
X=df[['age','anaemia','creatinine_phosphokinase','diabetes','ejection_fraction','high_blood_pressure','platelets','serum_creatinine','serum_sodium','sex','smoking','time']].values

In [5]:
y=df['DEATH_EVENT'].values

### Chi Square

In [19]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

In [24]:
bestfeatures=SelectKBest(score_func=chi2,k='all')
fiit=bestfeatures.fit(X,y)

In [25]:
dfscores=pd.DataFrame(fiit.scores_)
dfcolumns=pd.DataFrame(df.columns)

In [26]:
featureScores=pd.concat([dfcolumns,dfscores],axis=1)
featureScores.columns=['Specs','Score']

In [27]:
featureScores

Specs         Score
0                        age     44.619455
1                    anaemia      0.746593
2   creatinine_phosphokinase   1897.314839
3                   diabetes      0.000657
4          ejection_fraction     79.072541
5        high_blood_pressure      1.221539
6                  platelets  26135.771990
7           serum_creatinine     19.814118
8               serum_sodium      1.618175
9                        sex      0.001956
10                   smoking      0.032347
11                      time   3826.892661
12               DEATH_EVENT           NaN

### Correlation

In [6]:
from scipy import stats

In [7]:
Pearson_coef1,p_value1=stats.pearsonr(df['age'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(0.2537285430880036, 8.916762946531661e-06)

In [8]:
Pearson_coef1,p_value1=stats.pearsonr(df['anaemia'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(0.06627009846028778, 0.2532988145643253)

In [9]:
Pearson_coef1,p_value1=stats.pearsonr(df['creatinine_phosphokinase'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(0.06272816025237474, 0.27961124573610285)

In [10]:
Pearson_coef1,p_value1=stats.pearsonr(df['diabetes'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(-0.0019428833442034722, 0.9733118267846151)

In [11]:
Pearson_coef1,p_value1=stats.pearsonr(df['ejection_fraction'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(-0.26860331239406215, 2.4528974182084116e-06)

In [12]:
Pearson_coef1,p_value1=stats.pearsonr(df['high_blood_pressure'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(0.07935105769128536, 0.17114949852932787)

In [13]:
Pearson_coef1,p_value1=stats.pearsonr(df['serum_sodium'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(-0.19520359641640114, 0.0006889112363762405)

In [14]:
Pearson_coef1,p_value1=stats.pearsonr(df['sex'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(-0.0043163763197031, 0.9407519123205339)

In [16]:
Pearson_coef1,p_value1=stats.pearsonr(df['smoking'],df['DEATH_EVENT'])
Pearson_coef1,p_value1

(-0.012623152709359646, 0.8279207128091145)

## Feature Scaling

In [17]:
X=df[['age','anaemia','creatinine_phosphokinase','diabetes','ejection_fraction','high_blood_pressure','platelets','serum_creatinine','serum_sodium','sex','smoking','time']].values

In [19]:
y=df['DEATH_EVENT'].values

In [6]:
from sklearn import preprocessing
X=preprocessing.StandardScaler().fit(X).transform(X)

## Train Test Split

In [7]:
from sklearn.model_selection import train_test_split

In [8]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=4)

## Model Creation - Ensembling Algorithms

### 1. Bagging Techniques

#### a) Random Forest

In [9]:
from sklearn.ensemble import RandomForestClassifier
raf = RandomForestClassifier(n_estimators=90, max_depth=9, bootstrap='False',
                               max_features = 'sqrt',
                               #n_jobs=-1, 
                             verbose = 1, criterion='entropy')
raf.fit(X_train,y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.3s finished


RandomForestClassifier(bootstrap='False', class_weight=None,
            criterion='entropy', max_depth=9, max_features='sqrt',
            max_leaf_nodes=None, min_impurity_decrease=0.0,
            min_impurity_split=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=90, n_jobs=None, oob_score=False,
            random_state=None, verbose=1, warm_start=False)

In [10]:
pred1=raf.predict(X_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.0s finished


In [11]:
from sklearn import metrics
metrics.accuracy_score(y_test,pred1)

0.8555555555555555

#### Hyperparameter tuning

In [ ]:
from sklearn.model_selection import GridSearchCV

params={'max_features':['sqrt','auto'],'max_depth':[1,2,3,4,5,6,7,8,9,10],'verbose':[0,1],'bootstrap': ['True', 'False'],
       'n_estimators':[10,20,30,40,50,60,70,80,90,100],'criterion':['gini','entropy']}

gd1=GridSearchCV(estimator=raf,param_grid=params,cv=3)

gd1.fit(X_train,y_train)

gd1.best_params_

#### b) Extra Tree Classifier

In [12]:
from sklearn.ensemble import ExtraTreesClassifier

In [13]:
etc = ExtraTreesClassifier(n_estimators=100, random_state=0, max_depth=9,criterion='entropy',verbose=0,bootstrap='False')
etc.fit(X, y)

ExtraTreesClassifier(bootstrap='False', class_weight=None,
           criterion='entropy', max_depth=9, max_features='auto',
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           n_estimators=100, n_jobs=None, oob_score=False, random_state=0,
           verbose=0, warm_start=False)

In [14]:
pred2=etc.predict(X_test)

In [15]:
metrics.accuracy_score(y_test,pred2)

0.9555555555555556

### 2. Boosting Techniques

#### a) AdaBoost

In [16]:
from sklearn.ensemble import AdaBoostClassifier

In [17]:
ab=AdaBoostClassifier(base_estimator=raf,n_estimators=90,learning_rate=1)

In [18]:
ab.fit(X_train,y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.0s finished


AdaBoostClassifier(algorithm='SAMME.R',
          base_estimator=RandomForestClassifier(bootstrap='False', class_weight=None,
            criterion='entropy', max_depth=9, max_features='sqrt',
            max_leaf_nodes=None, min_impurity_decrease=0.0,
            min_impurity_split=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=90, n_jobs=None, oob_score=False,
            random_state=None, verbose=1, warm_start=False),
          learning_rate=1, n_estimators=90, random_state=None)

In [19]:
pred3=ab.predict(X_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:    0.0s finished


In [20]:
metrics.accuracy_score(y_test,pred3)

0.8555555555555555

#### b) Gradient Boosting

In [21]:
from sklearn.ensemble import GradientBoostingClassifier

In [22]:
gb=GradientBoostingClassifier(n_estimators=90,learning_rate=0.5,max_features=2,max_depth=2,random_state=0)

In [23]:
gb.fit(X_train,y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.5, loss='deviance', max_depth=2,
              max_features=2, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=90,
              n_iter_no_change=None, presort='auto', random_state=0,
              subsample=1.0, tol=0.0001, validation_fraction=0.1,
              verbose=0, warm_start=False)

In [24]:
pred4=gb.predict(X_test)

In [25]:
metrics.accuracy_score(y_test,pred4)

0.8333333333333334

#### c) Extreme Gradient Boosting

In [33]:
! pip install xgboost

     |████████████████████████████████| 148.9MB 119kB/s eta 0:00:01                     | 25.9MB 5.3MB/s eta 0:00:24MB/s eta 0:00:23     |████████████████                | 74.4MB 5.5MB/s eta 0:00:14     |████████████████▎               | 75.7MB 5.5MB/s eta 0:00:14     |████████████████▍               | 76.1MB 5.5MB/s eta 0:00:14     |████████████████▌               | 76.7MB 5.5MB/s eta 0:00:14     |████████████████▋               | 77.2MB 5.5MB/s eta 0:00:14     |████████████████▊               | 77.8MB 5.5MB/s eta 0:00:13██████▉       | 115.7MB 6.9MB/s eta 0:00:05███████       | 116.5MB 6.9MB/s eta 0:00:05


In [34]:
from xgboost import XGBClassifier

In [37]:
xgb=XGBClassifier(n_estimators=90,learning_rate=0.3,random_state=0,max_depth=6)

In [38]:
xgb.fit(X_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
       importance_type='gain', interaction_constraints='',
       learning_rate=0.3, max_delta_step=0, max_depth=6,
       min_child_weight=1, missing=nan, monotone_constraints='()',
       n_estimators=90, n_jobs=0, num_parallel_tree=1,
       objective='binary:logistic', random_state=0, reg_alpha=0,
       reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method='exact',
       validate_parameters=1, verbosity=None)

In [39]:
pred5=xgb.predict(X_test)

In [40]:
metrics.accuracy_score(y_test,pred5)

0.8333333333333334

#### Finalized Scores

##### Random Forest = 85%
##### Extra Tree Classifier = 95%
##### AdaBoost = 85%
##### Gradient Boosting = 83%
##### Extreme Gradient Boosting = 83%